File: Week 6/2. Max Flow/max_flow.py
from collections import defaultdict

# This class represents a directed graph 
# using adjacency matrix representation
class Graph:
    def __init__(self, graph):
        self.graph = graph
        self.row = len(graph)
        
    def bfs(self, source, sink, parent):
        visited = [False] * self.row
        queue = []
        
        # Mark the source node as visited and enqueue it
        queue.append(source)
        visited[source] = True
        
        while queue:
            u = queue.pop(0)
            
            # Get all adjacent vertices of the dequeued vertex u
            # If a adjacent has not been visited, then mark it
            # visited and enqueue it
            for v, capacity in enumerate(self.graph[u]):
                if visited[v] == False and capacity > 0:
                    queue.append(v)
                    visited[v] = True
                    parent[v] = u
                    
                    if v == sink:
                        return True
                    
        return False
    
    def edmondsKarp(self, source, sink):
        source -= 1
        sink -= 1
        parent = [-1] * self.row
        max_flow = 0
        
        # Augment the flow while there is path from source to sink
        while self.bfs(source, sink, parent):
            path_flow = float("Inf")
            v = sink
            
            # Find minimum residual capacity of the edges along the
            # path filled by BFS
            while v != source:
                path_flow = min(path_flow, self.graph[parent[v]][v])
                v = parent[v]
                
            max_flow += path_flow
            
            # Update residual capacities of the edges and reverse edges
            # along the path
            v = sink
            while v != source:
                u = parent[v]
                self.graph[u][v] -= path_flow
                self.graph[v][u] += path_flow
                v = parent[v]
                
        return max_flow     
        
n, m = map(int, input().split())
source, sink = map(int, input().split())
graph = [[0] * n for _ in range(n)]
for _ in range(m):
    u, v, capacity = map(int, input().split())
    graph[u - 1][v - 1] = capacity
    
g = Graph(graph)
print(g.edmondsKarp(source, sink))

File: Week 6/2. Max Flow/max_flow.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

class Graph
{
private:
    int numVertices;
    vector<vector<int>> graph;

public:
    Graph(int numVertices) : numVertices(numVertices)
    {
        graph.resize(numVertices, vector<int>(numVertices, 0));
    }

    void addEdge(int u, int v, int w)
    {
        graph[u - 1][v - 1] = w;
    }

    bool bfs(vector<vector<int>> &residualGraph, int source, int sink, vector<int> &parent)
    {
        vector<bool> visited(numVertices, false); // visited vertices
        queue<int> q;
        q.push(source);
        visited[source] = true;
        parent[source] = -1;

        while (!q.empty())
        {
            int u = q.front();
            q.pop();

            for (int v = 0; v < numVertices; ++v)
            {
                if (!visited[v] && residualGraph[u][v] > 0)
                {
                    if (v == sink)
                    {
                        parent[v] = u;
                        return true;
                    }

                    q.push(v);
                    parent[v] = u;
                    visited[v] = true;
                }
            }
        }

        return false;
    }

    int edmondsKarp(int source, int sink)
    {
        source--;
        sink--;
        // Create a residual graph and fill the residual graph
        // with given capacities in the original graph as
        // residual capacities in residual graph
        vector<vector<int>> residualGraph = graph; // residual graph
        vector<int> parent(numVertices);           // parent of each vertex

        int maxFlow = 0; // max flow

        // Augment flow while there is a path from source to sink
        while (bfs(residualGraph, source, sink, parent))
        {
            int pathFlow = INT_MAX; // flow through the path
            // Find minimum residual capacity of the edges along
            // the path filled by BFS, or find the maximum flow through the path found.
            for (int v = sink; v != source; v = parent[v])
            {
                int u = parent[v];
                pathFlow = min(pathFlow, residualGraph[u][v]);
            }

            // Update residual capacities of the edges
            // and reverse edges along the path
            for (int v = sink; v != source; v = parent[v])
            {
                int u = parent[v];
                residualGraph[u][v] -= pathFlow; // use the pathFlow units of its capacity in the residual graph
                residualGraph[v][u] += pathFlow; // add the pathFlow units of capacity to the reverse edge
            }

            // Add path flow to overall flow
            maxFlow += pathFlow;
        }

        return maxFlow;
    }
};

int main()
{
    int n, m;
    cin >> n >> m;

    int source, sink;
    cin >> source >> sink;

    Graph graph(n);

    for (int i = 0; i < m; ++i)
    {
        int u, v, w;
        cin >> u >> v >> w;
        graph.addEdge(u, v, w);
    }

    int maxFlow = graph.edmondsKarp(source, sink);
    cout << maxFlow << endl;
    return 0;
}